core/src/main/java/org/infinispan/interceptors/locking/ClusteringDependentLogic.java

```diff
@@ -476,15 +480,28 @@ public int getSegmentForKey(Object key) {
     */
    class ReplicationLogic extends InvalidationLogic {
       private StateTransferLock stateTransferLock;
+      private DistributionManager distributionManager;
 
       private final WriteSkewHelper.KeySpecificLogic localNodeIsPrimaryOwner = this::localNodeIsPrimaryOwner;
 
       @Inject
-      public void init(StateTransferLock stateTransferLock) {
+      public void init(StateTransferLock stateTransferLock, DistributionManager distributionManager) {
          this.stateTransferLock = stateTransferLock;
+         this.distributionManager = distributionManager;
+      }
+
+      @Override
+      public boolean localNodeIsOwner(Object key) {
+         return distributionManager.getDistributionInfo(key).write().isOwner();
       }
 
       @Override
+      public boolean localNodeIsPrimaryOwner(Object key) {
+         return distributionManager.getDistributionInfo(key).read() == Ownership.PRIMARY;
```

Comment:
I only meant I changed `DistributionInfo` so that it doesn't have 2 `Ownership` methods, each possibly being `PRIMARY`.